### ISSUE TITLE
Browser version?

### ISSUE THREAD

**Issue by fergardi** [aholstenson/miio#115]
Hello!

I'm really interested in your library. I'm brand new to NodeJS ecosystem, so maybe this is an stupid question, but I tried to use your library in a new web app I'm developing (with VueJS), just to discover I'm "missing" several packages, such as "fs" or "net" among others. I'm aware there are libraries meant for server side apps with NodeJS only. Is this one of these? It really needs to be server side? Cannot be used client side, considering it "only" needs to send/receive commands over the network, with a promise (even async/await) based mechanism? Is there any possibility to convert this library to a browser compatible one? Could I help you with that?

I'm just curious.

Thanks in advance.

**Comment by aholstenson** [comment]
Thanks for opening an issue! And sorry for the late reply.

Miio uses a lot of functionality that is only available in a Node-environment and not a browser-environment. The big thing being UDP networking which is not possible in a browser. One way to solve this is to bridge things over HTTP to a server that runs the library. Several users have done similar things in their own projects and setups.

I have a related project that I work on when I have time called [Tinkerhub](https://github.com/tinkerhub/tinkerhub) which is a "IoT container" in that it connects devices on different computers together in a shared environment and it has support for Miio via a plugin. Some time in the future Tinkerhub will have full support for running in a browser and connect to the local "thing network" via a HTTP server. When that support is finished that would allow for using the same API as Miio already exposes but in the browser, the request would then be sent to a HTTP server which would pass them on to the Miio-devices.

I don't know when that'll be finished though, I have way to many things going on right now, but hopefully within the next few months.

**Comment by fergardi** [comment]
Thanks for your reply.

Let's get us a step backwards. If I wanted to replicate the actual funcionality between the Xiaomi MiApp and one of my Xiaomi Yeelights without a server, could that be possible? I believe this is working due to a Xiaomi middle server between both, receiveing petitions and sending UDPs to my devices. But how does the Xiaomi Gateway works then? It acts as a standalone middle server too? Or it depends heavily on the Xiaomi server as usual?

TL;DR: I cannot replicate with my own Nodejs Browser side Android App the orders required to control my devices?
